include/aspect/particle/world.h

```diff
@@ -376,14 +369,19 @@ namespace aspect
                                  const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const;
 
         /**
-         * Finds the cells containing each particle for all particles. If
-         * particles moved out of this subdomain they will be sent
-         * to their new process and inserted there. After this function call
-         * every particle is either on its current process and in its current
-         * cell, or deleted (if it could not find its new process or cell).
+         * Finds the cells containing each particle for all particles in
+         * @p particles_to_sort. If particles moved out of the local subdomain
+         * they will be sent to their new process and inserted there.
+         * After this function call every particle is either on its current
+         * process and in its current cell, or deleted (if it could not find
+         * its new process or cell).
+         *
+         * @param [in,out] particles_to_sort Multimap containing all particles
+         * that will be sorted in this function. After the function call this
+         * container is empty.
          */
         void
-        sort_particles_in_subdomains_and_cells();
+        sort_particles_in_subdomains_and_cells(std::list<std::pair<types::LevelInd, Particle<dim> > > &particles_to_sort);
```

Comment:
This contains all particles that have left their cell during one advection step. An upper bound would be the total number of particles in the local domain, but that is likely a hundred times the usual size of this list. My thinking was that lists are mostly slow when *accessing* elements, while they are much faster when *removing* elements (and also *adding* elements when no reasonable upper bound for the vector is known, although I do not know if adding elements to a vector with known length is faster than adding elements to a list). All elements of the lists that I pass around here are only accessed twice, once when the new cell is determined, once when they are sorted according to the new cells. On the other hand a significant percentage of the total size is removed (all particles that need to be sent to other domains), or added (all particles that are received from other processes). 
To use a vector I would essentially need to pass a const input vector, and copy the whole vector element by element into a new one (except the elements that are removed). Do you think that would be faster? I am not sure, but could do a test (changes should be relatively small).